lib/spack/spack/test/util/compression.py

```diff
@@ -0,0 +1,174 @@
+        try:
+            import lzma  # noqa # novermin
+        except ImportError:
+            archiver = which('tar')
+            archiver.add_default_arg('-cfFoo.%s' % extension)
+            mode = 'system_tar'
+    out = os.path.join(ext_dir, 'Foo.%s' % extension)
+    if mode == 'tarfile':
+        archive = archiver.open(out, 'w%s' % compression)
+        archive.add(archive_file)
+        archive.close()
+    elif mode == 'system_tar':
+        with working_dir(ext_dir):
+            archiver(os.path.basename(archive_file))
+    else:
+        zip = archiver.ZipFile(out, 'w')
+        zip.write(archive_file)
+        zip.close()
+
+
+def derive_compression_algo(ext, style):
+    if ext and re.match(r'\.?zip$', ext):
+        return 'zipfile' if style == "native" else "unzip"
+    if ext and re.match(r'gz', ext):
+        return 'gzip' if style == "native" else "gzip"
+    if ext and re.match(r'bz2', ext):
+        return 'bz2' if style == "native" else "bunzip2"
+    if ext and not re.search(r'\.?Z', ext):
+        return 'tarfile' if style == "native" else "tar"
+
+
+def is_on_system(util):
+    try:
+        which(util, required=True)
+    except CommandNotFoundError:
+        return False
+    return True
+
+
+def module(ext, archiver=None):
+    try:
+        mod = __import__(ext)
+        if archiver:
+            archiver[0] = mod
+    except ImportError:
+        return False
+    return True
+
+
+def skip(extension, resource_type):
+    if resource_type == "native":
+        can_skip = module(derive_compression_algo(extension, resource_type))
+    else:
+        can_skip = is_on_system(derive_compression_algo(extension, resource_type))
+    return can_skip
+
+
+@pytest.fixture
+def archive_file(tmpdir_factory, request):
+    """ Creates a temporary directory containing a temp empty archive file"""
+    extension = request.param
+    mod = 'tarfile'
+    archiver = [0]
+    compression = ':' + extension.lower()
+    decomp_ext = extension.split('.')
+    if decomp_ext[1:]:
+        compression = ':' + decomp_ext[-1]
+    elif re.match('t', extension) and not re.match('tar', extension):
+        compression = ':' + extension[1:]
+        if extension[1:] == 'bz':
+            compression = compression + '2'
+    if extension == 'zip':
+        mod = 'zipfile'
+    if not module(mod, archiver):
+        pytest.skip('Archiver Module not available')
+    tmpdir = tmpdir_factory.mktemp('compression')
+    test = tmpdir.mkdir('test')
+    test_file = test / "comp.txt"
+    test_file.write("Spack Compression Test")
+    build_temp_archive(extension, mod, archiver[0], compression, str(test), str(tmpdir))
+    return os.path.join(str(tmpdir), 'Foo.%s' % extension)
+
+
+@pytest.mark.parametrize('archive_file', scomp.ALLOWED_ARCHIVE_TYPES, indirect=True)
+def test_native_unpacking(tmpdir_factory, archive_file):
+    extension = scomp.extension(archive_file)
+    if not skip(extension, "native"):
```

Comment:
I think it would be preferable to avoid dynamic test skipping: our CI has setup scripts which can install any needed tools, and in general we have tests that will fail without running that. It also avoids cases where those setup scripts fail or the compression algorithms are unavailable due to failures we should address.